evm/v0.5/contracts/PreCoordinator.sol

```diff
@@ -0,0 +1,417 @@
+   * @param _oracles The list of oracle contract addresses.
+   * @param _jobIds The corresponding list of Job IDs.
+   * @param _payments The corresponding list of payment amounts.
+   */
+  function createServiceAgreement(
+    uint256 _totalPayment,
+    uint256 _minResponses,
+    address[] calldata _oracles,
+    bytes32[] calldata _jobIds,
+    uint256[] calldata _payments
+  )
+    external onlyOwner returns (bytes32 saId)
+  {
+    require(_oracles.length == _jobIds.length && _oracles.length == _payments.length, "Unmet length");
+    require(_oracles.length <= MAX_ORACLE_COUNT, "Cannot have more than 45 oracles");
+    require(_oracles.length >= _minResponses, "Invalid min responses");
+    saId = keccak256(abi.encodePacked(globalNonce, now));
+    globalNonce++; // yes, let it overflow
+    // Manually calculate total payment off-chain
+    serviceAgreements[saId] = ServiceAgreement(_totalPayment, _minResponses, _oracles, _jobIds, _payments);
+
+    emit NewServiceAgreement(saId, _totalPayment, _minResponses);
+  }
+
+  /**
+   * @notice This is a helper function to retrieve the details of a service agreement
+   * by its given service agreement ID.
+   * @dev This function is used instead of the public mapping to return the values
+   * of the arrays: oracles, jobIds, and payments.
+   */
+  function getServiceAgreement(bytes32 _saId)
+    external view returns
+  (
+    uint256 totalPayment,
+    uint256 minResponses,
+    address[] memory oracles,
+    bytes32[] memory jobIds,
+    uint256[] memory payments
+  )
+  {
+    return
+    (
+      serviceAgreements[_saId].totalPayment,
+      serviceAgreements[_saId].minResponses,
+      serviceAgreements[_saId].oracles,
+      serviceAgreements[_saId].jobIds,
+      serviceAgreements[_saId].payments
+    );
+  }
+
+  /**
+   * @notice Deletes a service agreement from storage
+   * @dev Use this with caution since there may be responses waiting to come
+   * back for a service agreement. This can be monitored off-chain by looking
+   * for the ServiceAgreementRequest event.
+   * @param _saId The service agreement ID
+   */
+  function deleteServiceAgreement(bytes32 _saId) external onlyOwner {
+    delete serviceAgreements[_saId];
+  }
+
+  /**
+   * @notice Returns the address of the LINK token
+   * @dev This is the public implementation for chainlinkTokenAddress, which is
+   * an internal method of the ChainlinkClient contract
+   */
+  function getChainlinkToken() external view returns (address) {
+    return chainlinkTokenAddress();
+  }
+
+  /**
+   * @notice Called when LINK is sent to the contract via `transferAndCall`
+   * @dev The data payload's first 2 words will be overwritten by the `_sender` and `_amount`
+   * values to ensure correctness. Calls oracleRequest.
+   * @param _sender Address of the sender
+   * @param _amount Amount of LINK sent (specified in wei)
+   * @param _data Payload of the transaction
+   */
+  function onTokenTransfer(
```

Comment:
So for this, I got the PreCoordinator contract on the LinkTokenReceiver, but the Oracle contract will revert when using `transferAndCall` to it. Unfortunately debugging that is difficult because the source of LinkToken isn't in the project. So maybe that will be best for a separate PR.

Reply:
Got this resolved.